File: Python/9012.py
import sys

# VPS 확인
def VPS(string):
    cnt = 0
    for i in string:
        if i == '(':
            cnt += 1
        elif i == ')':
            cnt -= 1
        if cnt < 0:
            return("NO")

    if cnt == 0:
        return("YES")
    else:
        return("NO")

# 반복 횟수 입력
n = int(sys.stdin.readline().rstrip())

# n번 반복
for i in range(n):
    # 괄호 입력받기
    string = sys.stdin.readline().rstrip()
    print(VPS(string))

File: Java/_9012.java
package Baekjoon;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;

public class _9012 {

    public static void main(String[] args){

        try {

            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            int n = Integer.parseInt(br.readLine());

            for(int i = 0; i < n; i++){
                String str = br.readLine();
                System.out.println(VPS(str.trim()));
            }

        } catch (Exception e) { e.printStackTrace(); }
    }

    public static String VPS(String str){

        Stack<Character> stack = new Stack<Character>();

        for(int i = 0; i < str.length(); i++){
            if(str.charAt(i) == '('){
                stack.push(str.charAt(i));
            }
            else{
                if(!stack.isEmpty()){
                    stack.pop();
                }
                else{
                    return("NO");
                }
            }
        }

        if(stack.isEmpty()){
            return("YES");
        }
        else{
            return("NO");
        }
    }
}
